File: src/harness/logger.py
from __future__ import annotations

import json
from dataclasses import asdict
from datetime import datetime
from pathlib import Path

from .types import TurnRecord


class TrajectoryLogger:
    def __init__(self, log_dir: str | Path) -> None:
        log_path = Path(log_dir)
        log_path.mkdir(parents=True, exist_ok=True)
        stamp = datetime.now().strftime("%Y%m%d-%H%M%S-%f")
        self.path = log_path / f"trajectory-{stamp}.jsonl"

    def append(self, record: TurnRecord) -> None:
        with self.path.open("a", encoding="utf-8") as f:
            f.write(json.dumps(asdict(record), ensure_ascii=False) + "\n")

File: ts/src/logger.ts
/**
 * Trajectory logger — JSONL append-only log of turn records.
 */

import { mkdirSync, appendFileSync } from "node:fs";
import { join } from "node:path";
import type { TurnRecord } from "./types.js";

export class TrajectoryLogger {
  public readonly path: string;

  constructor(logDir: string) {
    mkdirSync(logDir, { recursive: true });

    const now = new Date();
    const pad = (n: number, w = 2) => String(n).padStart(w, "0");
    const stamp = [
      now.getFullYear(),
      pad(now.getMonth() + 1),
      pad(now.getDate()),
      "-",
      pad(now.getHours()),
      pad(now.getMinutes()),
      pad(now.getSeconds()),
      "-",
      pad(now.getMilliseconds() * 1000, 6), // microsecond-ish resolution
    ].join("");

    this.path = join(logDir, `trajectory-${stamp}.jsonl`);
  }

  /** Append a single turn record as one JSON line. */
  append(record: TurnRecord): void {
    const line = JSON.stringify(record) + "\n";
    appendFileSync(this.path, line, "utf-8");
  }
}
